### PR TITLE
Added DP Solution for Optimal BST Problem

### PR DESCRIPTION
**Pull request by Marethyu12**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [ ] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by cclauss** [comment]
What does bst stand for and why should the reader have to guess?

**Comment by Marethyu12** [comment]
oops, ty for pointing out. I fixed it.

**Comment by cclauss** [comment]
Let's move this to the `searches` directory and rename it to `searches/binary_search_tree.py`.  Then we need a bit of a description of how this is different from / better than `searches/binary_search.py`

File: dynamic_programming/optimal_bst.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/python3
```
Comment (cclauss) [inline dynamic_programming/optimal_bst.py]: #!/usr/bin/env python3

File: dynamic_programming/optimal_bst.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# tree while the nodes with low frequencies will be placed near the tree
+# leaves thus reducing the search time.
+
+import sys
+
+from random import randint
+
+
+class Node:
+    """BST Node"""
```
Comment (cclauss) [inline dynamic_programming/optimal_bst.py]: """Binary Search Tree Node"""

File: dynamic_programming/optimal_bst.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    print_binary_search_tree(
+        root, key, root[i][j] + 1, j, key[root[i][j]], False
+    )  # recur to right child
+
+
+def find_optimal_binary_search_tree(nodes):
+    """
+    This function calculates and prints the optimal BST.
+    The dynamic programming algorithm below runs in O(n^2) time.
+    Implemented from CLRS book.
```
Comment (cclauss) [inline dynamic_programming/optimal_bst.py]: Expand the acronym CLRS.

File: dynamic_programming/optimal_bst.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    
+    >>> find_optimal_binary_search_tree([Node(12, 8), Node(10, 34), Node(20, 50), Node(42, 3), Node(25, 40), Node(37, 30)])
```
Comment (cclauss) [inline dynamic_programming/optimal_bst.py]: Wrap all lines to 88 chars max.

File: dynamic_programming/optimal_bst.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    # Tree nodes must be sorted first, the code below sorts the keys in
+    # increasing order and rearrange its frequencies accordingly.
+    nodes.sort(key=lambda node: node.key)
+
+    n = len(nodes)
+
+    key = [nodes[i].key for i in range(n)]
+    freq = [nodes[i].freq for i in range(n)]
+
+    # This 2D array stores the overall tree cost (which's as minimized as possible); for a single key, cost is equal to frequency of the key.
```
Comment (cclauss) [inline dynamic_programming/optimal_bst.py]: Wrap all lines to 88 chars max.

File: dynamic_programming/optimal_bst.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    sum = [[freq[i] if i == j else 0 for j in range(n)] for i in range(n)]
+    # stores tree roots used for constructing BST later
+    root = [[i if i == j else 0 for j in range(n)] for i in range(n)]
+
+    for l in range(2, n + 1):  # l is an interval length
+        for i in range(n - l + 1):
+            j = i + l - 1
+
+            dp[i][j] = sys.maxsize  # set the value to "infinity"
+            sum[i][j] = sum[i][j - 1] + freq[j]  # (sum in range [i...j]) = (sum in range [i...j - 1]) + freq[j]
```
Comment (cclauss) [inline dynamic_programming/optimal_bst.py]: Move the comment to the line above or eliminate it.

**Comment by Marethyu12** [comment]
I followed your requests except one where you asked me to move this file to searches directory and rename it. I rejected this request because my original intent is to add a new dynamic programming algorithm not to make a general BST (it can be done by someone else). It's already obvious that I'm focusing on specific class of BST namely optimal BSTs.

**Comment by Marethyu12** [comment]
Plus binary_search_tree.py already exist in data structures directory.

**Pull request merged**